### ISSUE TITLE
What is the correct way to define custom pydantic types that also play nicely with JSON Schema?

### ISSUE THREAD

**Issue by calvinwyoung** [tiangolo/fastapi#1285]
### First check

* [X] I used the GitHub search to find a similar issue and didn't find it.
* [X] I searched the FastAPI documentation, with the integrated search.
* [X] I already searched in Google "How to X in FastAPI" and didn't find any information.

### Description

What's the best way to define a custom data type that works with both pydantic and JSON Schema?

To illustrate, let's say we want to implement a custom `Arrow` datetime type (this was also asked in #1186). Borrowing the approach from #452, we could implement the following:

```python
class Arrow(datetime.datetime):
    @classmethod
    def __get_validators__(cls):
        yield cls.validate

    @classmethod
    def validate(cls, v):
        return arrow.get(v)
```

This is kind of weird for 2 reasons:

1. We have to inherit from `datetime.datetime` in order for JSON Schema to recognize that this should be represented as a datetime-formatted string. This isn't entirely intuitive — someone could reasonably wonder why we're inherting from `datetime.datetime`, but then returning an `Arrow` object in `validate`.

2. This only lets us define how we want to *decode* from JSON. To handle encoding, I have to use `json_encoders` in my model:

```python
class TestModel(pydantic.BaseModel):
    datetime: Arrow

    class Config:
        json_encoders = {arrow.Arrow: lambda obj: obj.isoformat()}
```

Notice how the `json_encoders` entry references an `arrow.Arrow` object, and not the `Arrow` type that was defined earlier. This also seems suboptimal.

----------------------

Is this the correct way to implement custom types? Or is there a better solution?

Ideally I'd love to be able to be able to define a custom type similar to the way `marshmallow` handles it (i.e., by overriding `_serialize` and `_deserialize` methods):
https://marshmallow.readthedocs.io/en/stable/custom_fields.html

### Additional context

I recognize that the second point is more closely related to pydantic than it is to fastapi, and certainly there's been some discussion about it (see https://github.com/samuelcolvin/pydantic/issues/951). But that still doesn't address the first point around playing nicely with JSON schema.

I'm kind hoping fastapi can do something on top of pydantic to make this all a bit smoother.

**Comment by tiangolo** [comment]
Yeah, having types declare their own schema and their own serialization is an interesting idea. :sparkles: 

But it's actually all on the Pydantic side. JSON Schema is also generated by Pydantic, FastAPI gets the JSON Schemas from Pydantic models directly.

I think the first point would be the issue you linked in Pydantic. And then the next feature request would be to be able to generate a JSON Schema from the type itself.

But anyway, it would actually all be in Pydantic, not really in FastAPI...

**Comment by calvinwyoung** [comment]
@tiangolo That makes sense, thanks for the clarification! I'll mark this as resolved
